src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -825,95 +825,134 @@ void makeMemberMaybeNull(MethodSymbol method, string memberName)
             {
-                if (!state.Reachable)
+                var slot = GetOrCreateSlot(parameter);
+                if (slot <= 0)
                 {
-                    return;
+                    continue;
                 }
 
-                foreach (var parameter in this.MethodParameters)
+                var annotations = parameter.FlowAnalysisAnnotations;
+                var hasNotNull = (annotations & FlowAnalysisAnnotations.NotNull) == FlowAnalysisAnnotations.NotNull;
+                var parameterState = state[slot];
+                if (hasNotNull && parameterState.MayBeNull())
                 {
-                    var slot = GetOrCreateSlot(parameter);
-                    if (slot <= 0)
+                    Location location;
+                    if (syntaxOpt is BlockSyntax blockSyntax)
                     {
-                        continue;
-                    }
-
-                    var annotations = parameter.FlowAnalysisAnnotations;
-                    var hasNotNull = (annotations & FlowAnalysisAnnotations.NotNull) == FlowAnalysisAnnotations.NotNull;
-                    var parameterState = state[slot];
-                    if (hasNotNull && parameterState.MayBeNull())
-                    {
-                        // Parameter '{name}' must have a non-null value when exiting.
-                        Diagnostics.Add(ErrorCode.WRN_ParameterDisallowsNull, syntaxOpt?.GetLocation() ?? methodMainNode.Syntax.GetLastToken().GetLocation(), parameter.Name);
+                        location = blockSyntax.CloseBraceToken.GetLocation();
                     }
                     else
                     {
-                        EnforceNotNullIfNotNull(syntaxOpt, state, this.MethodParameters, parameter.NotNullIfParameterNotNull, parameterState, parameter);
+                        location = syntaxOpt?.GetLocation() ?? methodMainNode.Syntax.GetLastToken().GetLocation();
                     }
+
+                    // Parameter '{name}' must have a non-null value when exiting.
+                    Diagnostics.Add(ErrorCode.WRN_ParameterDisallowsNull, location, parameter.Name);
                 }
+                else
+                {
+                    EnforceNotNullIfNotNull(syntaxOpt, state, this.MethodParameters, parameter.NotNullIfParameterNotNull, parameterState, parameter);
+                }
+            }
+        }
+
+        void EnforceParameterNotNullWhenOnExit(SyntaxNode syntax, ImmutableArray<ParameterSymbol> parameters, bool sense, LocalState stateWhen)
+        {
+            if (!stateWhen.Reachable)
+            {
+                return;
+            }
+
+            foreach (var parameter in parameters)
+            {
+                ReportParameterIfBadConditionalState(syntax, parameter, sense, stateWhen);
             }
+        }
+
+        void EnforceNotNullWhenForPendingReturn(PendingBranch pendingReturn, BoundReturnStatement returnStatement)
```

Comment:
>EnforceNotNullWhenForPendingReturn [](http://example.com/codeflow?start=13&length=34)

Consider moving this before `EnforceParameterNotNullOnExit()` since it looks like that might reduce the differences in this section.